KeyControl: fix keylock/unlock bugs, reset pitch_adjust

ronso0 opened the pull request:
Fixes unintented pitch shift when un/locking repeatedly with "lock original key" and "unlock, keep current key", see https://github.com/mixxxdj/mixxx/pull/4710#issuecomment-1088142652 by resetting `pitch_adjust` when unlocking to linear pitch and locking to original key.

- [x] adjust pitchRoundtrip test
- [x] add test for "Unlock, keep current key" (or include it in pitchRoundtrip test)

ronso0 commented:
back to draft because I am also looking into another issue with key un/lock and out-of-bounds rate changes (pitch/key can not be reset easily)

ywwg commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -94,6 +94,10 @@ KeyControl::KeyControl(const QString& group,
+    // Internal m_pitchRateInfo.keylock is updated only in updateRate() when any
+    // rate, keylock or vc controls are touched. Though, it may be requested by
+    // EngineBuffer, e.g when nudging a track. Thus, update it now.
+    m_pitchRateInfo.keylock = m_pKeylock->get();
```
oo I like the inline clazy notes, very fancy

daschuer replied:
```suggestion
    m_pitchRateInfo.keylock = m_pKeylock->toBool();
```

ronso0 replied:
clazy?
crazy & lazy

ronso0 replied:
oh you mean the CI
already fixed

ronso0 commented:
The other issue I'm attempting to fix is this:
When I lock and unlock the key often, with rate changes in between, during a set my `pitch != 0.0` indicator flashes even though I just restored the original key. Now I finally managed to reproduce this with a script, and also manually with the GUI `+`/`-` buttons. Appearantly the crucial part is to move the rate slider outside the GUI/controller range.
* Preferences:
  - [x] Keylock: lock current key
  - [x] Keyunlock: keep current key
  * rate range: 8%, slider down increases rate
* Procedure:
  * deck1 keylock OFF, rate centered, no prior rate or key adjustments on that deck!
  * load track
  * move rate slider out of bounds (via Sync or scripting `engine.setValue(group, "rate", 1.1)`)
  * lock key
  * reset key to original with `reset_key` (pitch = 0.0)
  * move rate slider a bit
  * unlock key (pitch should still be `0.0`)
  * pitch is something like 1,735221e-17

See [0a6d00d](https://github.com/mixxxdj/mixxx/pull/4710/commits/0a6d00d608080921a2cffb0a2953c279efa16713)
I don't know how to avoid the calculation round-trip of `m_pitchRateInfo` members in the first place without making `updateRate()` even more complex. Hope this solution is okay.

ronso0 commented:
_ping._ Any more comments?

ronso0 commented:
I've just split the last commit into two to make it easier to review.

I'll also manual testing again with all 4 keylock/unlock combinations.

Swiftb0y commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -256,6 +256,13 @@ void KeyControl::updateRate() {
     qDebug() << "      =" << qSetRealNumberPrecision(18) << m_pitchRateInfo.pitchRatio;
     qDebug() << "      =" << m_pitchRateInfo.pitchRatio;
 #endif
+    // If we just unlocked the original key with speed != +-1.0 we may encounter wrong
+    // decimals (e.g. 1 - 1.73123e-09) after pitchRatio calculation round-trip.
+    // Round to 1.0 to avoid false positive pitch offset (pitchOctaves != 0).
+    double pitchRatioDiffTo1 = fabs(1.0 - m_pitchRateInfo.pitchRatio);
+    if (0.0 < pitchRatioDiffTo1 && pitchRatioDiffTo1 < pow(10, -9)) { // 0.000000001
+        m_pitchRateInfo.pitchRatio = 1.0;
+    }
```
I'm struggling to understand the reasoning behind this because I struggle to understand the surrounding code. `m_pitchRateInfo.pitchRatio` is basically imprecise because of rounding errors? But only around 1.0?

ronso0 replied:
Basically roundtrip rounding erros, yes.
Tbh I didn't manage to understand how all the ratios relate in all situations either because I'm still confused by the naming, but with debug output enabled it's at least easy to trace how the `m_pitchRateInfo` values change. In https://github.com/mixxxdj/mixxx/pull/4710#issuecomment-1088142652 I outlined the circumstances and the procedure to reproduce the imprecision.

daschuer replied:
The issue is that the new statement in line 216 is designed to make m_pitchRateInfo.pitchRatio = 1, but it is done indirect via other variables. It would be straight forward and more easy to understand to set it 1 explicit. This would require more changes in the code though.

daschuer commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -11,7 +11,8 @@
 #include "track/keyutils.h"
 
-//static const double kLockOriginalKey = 0;
+#define ENABLE_DEBUG_OUTPUT false
+
```
The difference is that the code in #if is ignored by the compiler, the code in "if constexpr" is checked and optimized away later.

Swiftb0y replied:
True, but it doesn't make much of a difference in terms of what ends up in the binary... Metaprogramming using the C++ preprocessor is discouraged from (in C++) AFAIK.

ronso0 replied:
Alright, thanks!

JoergAtGithub replied:
The preprocessor variant is not type safe. But the preprocessor has the benefit, that you can define these macros from outside (compiler commandline call). But here the define is in the source file.

ronso0 commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -137,24 +138,25 @@ void KeyControl::slotRateChanged() {
+    qDebug() << ".";
+    qDebug() << "   KeyControl::updateRate";
+    qDebug() << "      |  tempoRatio" << qSetRealNumberPrecision(18) << m_pitchRateInfo.tempoRatio;
+#endif
```
oki, fixed.

ronso0 commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -333,7 +409,19 @@ void KeyControl::slotSyncKey(double v) {
+        // If we reset the key while the rate is centered and keylock is enabled
+        // pitchTweakRatio would not be reset automatically.
+        // Reset m_pitchRateInfo members manually
+        if (m_pitchRateInfo.tempoRatio == 1.0) {
```
Unfortunately some time passed and I don't remember anymore how to reproduce the issue.
Members are reset correctly if I put only the debug commit on top of 2.3

By now I'd rather start over and
* add the debug output in the suggested way
* [update m_pitchRateInfo.keylock in init](https://github.com/mixxxdj/mixxx/pull/4710/commits/889cb15dc6b271bb397da283c96ba74c36771441)
* apply the rounding issue compensation commits
* drop the key reset commit

The functional changes are quite small so I hope this is okay for you?

daschuer replied:
Yes, go ahead.

ronso0 commented:
Uff, there is another bug:
(best experienced with src/test/sine-30.wav : )
* keylock: lock original key
* keyunlock: keep current key
* keylock disabled, enable repeat for continous stream 
1. play, change rate to like +4%, **no manual key change**
2. lock key => key reset to default, pitch = 0.0
3. unlock key => no key change, as expected with `keep current key`
4. lock key => key lowered, should reset to original key + pitch slider offset
5. repeating 3. and 4. will lower the key step by step

I'll go with
> It would be straight forward and more easy to understand to set it 1 explicit. This would require more changes in the code though.

in order to rebase this whole thing.

ronso0 commented:
> Uff, there is another bug: (best experienced with src/test/sine-30.wav : )
> 
> * keylock: lock original key
> * keyunlock: keep current key
> * keylock disabled, enable repeat for continous stream
> 
> 1. play, change rate to like +4%, **no manual key change**
> 2. lock key => key reset to default, pitch = 0.0
> 3. unlock key => no key change, as expected with `keep current key`
> 4. lock key => key lowered, should reset to original key + pitch slider offset
> 5. repeating 3. and 4. will lower the key step by step

When I fix that bug by setting both `pitchRatio` and `pitchTweakRatio` manually for each keylock/unlock and locked/unlocked case...
all manual tests succeed, Mixxx behaves as expected.
BUT the [Pitchroundtrip test](https://github.com/mixxxdj/mixxx/blob/0ad01b6ccc744045fb727eb4e775aa0e52bf4998/src/test/enginebuffertest.cpp#L53) fails here:
```
Note: Google Test filter = EngineBufferTest.PitchRoundtrip
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EngineBufferTest
[ RUN      ] EngineBufferTest.PitchRoundtrip
/home/ronso/Downloads/mixxx_source/src/master/src/test/enginebuffertest.cpp:75: Failure
Expected equality of these values:
  0.5
  ControlObject::get(ConfigKey(m_sGroup1, "pitch"))
    Which is: 0
``` 
The test assumes that `pitch_adjust` remains at x.y even if the original key is locked.

@daschuer Unfortunately some rules aren't clear to me from the code comments / manual ([pitch_adjust](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-pitch_adjust)) and I'd appreciate if you could clarify:
How is the `pitch_adjust` knob supposed to behave?
If we assume that user have that mapped to a rotary knob, should the hardware knob _always_ stay in sync with the internal state **even when magic buttons are pressed**?
Unfotunately the current behaviour is inconsistent:
* the knob is **static** when I "lock at original pitch" after `pitch_adjust`ments (though it doesn't reset to original key, for that  I'd need to reset the adjust knob, too)
* the knob **moves** when I "unlock, keep current key"
* the knob **moves** when I press "match key" or "reset key"

This is very confusing.
Should _keylock/unlock_ be treated differently than _match/reset_?

To be clear, in this stable banch I don't necessarily want to refactor `updateRate` to fix that bug **and** make the test still pass / meet established user expectations.

I don't want to open a can of worms and refactor the stable branch just to fix that bug (and make tests still pass).
> hardware knob _always_ stay in sync with the internal state

ronso0 committed:
KeyControl: simplify updatePitch() & updatePitchAdjust(), more const

daschuer commented:
I have just tested with the 2.3 and the control is implemented on top of keylock. It is not changed if the pitch is adjusted by keylock. For my understanding this behavior is desired, because you will touch the control likely to adjust the pitch from what you have currently. 
Scenario: 
* Adjust the tempo from matching speed 
* Hit Keylock to avoid chipmunk voices 
* Adjust the pitch for a matching key. 
First snap the control from soft-takeover and then notice that the range of +-3 Semitones has already been reached would feel cumbersome. Soft-Takover is required in mapping though, because the control is reset after loading new tracks.

The behavior is described here:  
https://github.com/mixxxdj/mixxx/blob/74ec3459356c60c74881ffb10371cfdd623f1cd4/src/engine/enginebuffer.cpp#L937
and here
https://github.com/mixxxdj/mixxx/blob/eff27abd571ac3e1b273ad4aef6407eceee0cfee/src/engine/controls/keycontrol.cpp#L37

The manual is not clear and can make use of an update: 
> Adjust the pitch in addition to the :term:`tempo` slider pitch.

daschuer commented:
Here is the manual update: https://github.com/mixxxdj/manual/pull/480

ronso0 commented:
> > * the knob is static when I "lock at original pitch" after pitch_adjustments (though it doesn't reset to original key, for that I'd need to reset the adjust knob, too)
> 
> That can be expected, since the pitch_adjust is applied to all other. Did you expect that the knob is reset and goes out of sync, in such situation?

Yes, I'd expect that that the pitch is reset to default, exactly as labeled in the preferences: Keylock mode "Original key". That would be an easy way to reset the key if you don't have `reset_key` mapped, for example after having matched the key.
If you have `pitch_adjust mapped` you'd then have access to the full +-3 semitones range.

Also, this fixes the key shift bug with "lock original" + "unlock keep current".

> The "unlock, keep current key" mode is part of an "allow toggling keylock during playback" use case. It allows to adjust the tempo without the pitch, and without a pitch jump during enabling. Not sure if this is used at all in the wild.

**I** use it : ) The motivation to implement it was that I have keylock enabled pretty much all the time, but sometimes I want to produce smooth pitchbend effects just with the jog wheel (or with the tempo slider). Then lock again, if necessary re-sync the beats with the jogwheel.
In conjunction with "Lock current key" this prevents key jumps when toggling the lock. I have `reset_key` mapped, as well as `pitch_adjustup/down` and never ran into issues (except the rounding issue I'm trying to fix here).

daschuer commented:
> Yes, I'd expect that that the pitch is reset to default,

OK, I have no objection to this change. It may actually feel more natural, even though it is a point release of a mature branch. Can we still consider it as part of a bugfix? 

> I use it : ) 

Great to hear. It this an unique selling point of Mixxx? Maybe a topic of a future news post ;-)

ronso0 commented:
> Great to hear. It this an unique selling point of Mixxx? Maybe a topic of a future news post ;-)

I started DJing on Linux so don't and didn't use any other soft than Mixxx : )
I just found Traktor users asking to revert the "Engaging Key Lock, resets a potential key offset" "fix"
[Disable "Engaging Key Lock, resets a potential key offset"](https://www.native-instruments.com/forum/threads/disable-engaging-key-lock-resets-a-potential-key-offset.358605/)
Serato, Rekordbox, idk
Traktor Pro 3 appears to have separate MIDI controls: Key Lock and Key Lock (Preserve Pitch), though no global preferences option for that :shrug: So at least Mixxx rules out Trakor in that regard since we have "Lock original" and "Keep current key" since #1222 : )

> Can we still consider it as part of a bugfix?

It fixes the continuous key shift when toggling keylock repeatedly with "Lock original" and "Unlock, keep current".
So I'd say yes.
> even though it is a point release of a mature branch.

I think only users who have `pitch_adjust knob` mapped to a regular rotary knob¹ get a minor benefit from the current behaviour as that knob is still in sync even after a sudden key change. Those who don't have it will have to resort to `key_reset`.
¹there is only one official mapping with that

So I think it's the same situation as with the crossfader reset after AutoDJ runs out:
users may take it as it is, but for others the current behaviour is an issue.

ronso0 committed:
KeyControl: (fix) make "Lock original key" also reset pitch_adjust

This prevents the continuous key shift when un/locking repeatedly with
"Lock original key" and "Unlock, keep current key".

ronso0 committed:
KeyControl: (fix) apply pitch_adjust on "Unlock, reset to linear pitch"

ronso0 commented:
I noticed another bug (and I have no idea why we/I implemented it that way in #1222):
key unlocked, rate off center, apply some pitch_adjust (match_key), disable keylock "Reset to linear pitch"
= pitchTweakRatio is reset but not applied to `pitch_adjust`
= next time touching that knob pitch would jump

ronso0 commented:
Furthermore I noticed the pitch/keylock tests don't consider the unlock mode, thus they are partially pointless (assuming the default unlock mode).
I will adjust the PitchRoundtrip test and write a new one to test the "Keep current" unlock mode.

ronso0 commented:
Considering I started this to fix a small initialization bug and minor anoyance with _my_ use case this is growing rather big...

ronso0 commented:
I think I will split this up:
* the first few commits (init keylock from config, restore pitchRatio precision)
* the other bugs incl. adjusted tests

So the simple fixes can go to the next bugfix release asap.

daschuer commented:
Yes, I did also consider this. This way we can get out 2.3.2 in de upcoming week or such.

ronso0 commented:
2.3.3, yes.

#4756

ronso0 committed:
DisableKeylockResetsPitch test: use KeyunlockMode::ResetLockedKey explicitly

ronso0 committed:
PitchRoundtrip test: (fix) use KeyunlockMode::ResetLockedKey explicitly

ronso0 committed:
PitchRoundtrip test: adjust to new pitch_adjust reset with key un/lock modes

ronso0 commented:
Alrigth, so I wrapped my head around this stuff again, finished it and I adjusted the PitchRoundTrip test accordingly.
Rebased onto 2.4

Only ToDo left is
- [ ] add test for "Unlock, keep current key" (or include it in pitchRoundtrip test)

ronso0 commented:
Reminder: to simplify testing I added key lock & unlock mode toggles, `pitch` and `pitch_adjust` sliders to LateNight.
Need to remove that commit before merge!

daschuer commented:
Oh, sorry this has stalled. Can you resolve it the conflicts? 
I will test and review it again.

daschuer commented:
The conflicts are only in the testing skin. 
I have just given this a short spin and I did not discover any issues. So you may want to put this is not a merge able state.

daschuer commented:
Is the track load behavior still in the desired state? 
I am unsue what is desired in the middle ground of not putting sliders out if sync but also start with a new track without supprises. Please verify everything matches your use case.

ronso0 committed:
Keylock: add test for KeyunlockMode::KeepLockedKey

ronso0 committed:
Merge remote-tracking branch 'mixxx/2.4' into keylock-fix

ronso0 commented:
Everything still looks okay when testing with my script.
I'll give this a more thorough spin next weekend with my controller.

I added a simple test for KeepLockedKey mode and removed the skin commit.

ronso0 commented:
> I am unsue what is desired in the middle ground of not putting sliders out if sync but also start with a new track without supprises.

I'm not that much into what I did here after my last edit months ago X), please give my some steps to reproduce / understand what you mean.

ronso0 commented:
```
/home/runner/work/mixxx/mixxx/src/test/enginebuffertest.cpp:62: Failure
Expected equality of these values:
  0.5
  ControlObject::get(ConfigKey(m_sGroup1, "pitch"))
    Which is: 0.5
```
??
Test passes only macOS ARM

ronso0 commented:
> Everything still looks okay when testing with my script.

In a manual test though I ended up with pitch_adjust != 0 again (couldn't reproduce until now, but I didn't do many runs):
* Lock: keep current
* Unlock: keep locked key

* raise pitch to 10% (outside pitch slider range)
* lock, unlock, do manual key ajustments etc.
* reset rate
* reset key
* = pitch_adjust != 0, just slightly off
* reset pitch_adjust slider = pitch slightly off
* reset pitch slider = pitch_adjust slightly off
* etc.
* = no way to completely reset the pitch controls (except LOCK, reset key, UN-lock)
:shrug: 

Will see if I can reproduce with my controller.

ronso0 commented:
> ```
> /home/runner/work/mixxx/mixxx/src/test/enginebuffertest.cpp:62: Failure
> Expected equality of these values:
>   0.5
>   ControlObject::get(ConfigKey(m_sGroup1, "pitch"))
>     Which is: 0.5
> ```
> 
> ?? Test passes only on macOS ARM

Anyone has a clue what's going on?

Holzhaus commented on src/test/enginebuffertest.cpp:
```diff
@@ -43,6 +45,23 @@ TEST_F(EngineBufferTest, DisableKeylockResetsPitch) {
+    ControlObject::set(ConfigKey(m_sGroup1, "keylock"), 0.0);
+    // We require a buffer process to see that the keylock state has changed.
+    ProcessBuffer();
+    ASSERT_EQ(0.5, ControlObject::get(ConfigKey(m_sGroup1, "pitch")));
```
```suggestion
    ASSERT_DOUBLE_EQ(0.5, ControlObject::get(ConfigKey(m_sGroup1, "pitch")));
```

Swiftb0y replied:
0.5 can be represented unambiguously though. So there is a rounding/precision issue somewhere that introduces a slight difference... possibly a bug?

ronso0 replied:
There definitely is a round-trip rounding issue, see https://github.com/mixxxdj/mixxx/blob/53f3b073bf19d0af7e745db565b4d1bcf2df6d3e/src/engine/controls/keycontrol.cpp#L276-L281
and #12027 but fixing that is out of scope for this PR IMO.
I suspect it's `KeyControl::updateKeyCOs` recalculating all values, even pitch if the called by `updatePitch()` which is called by `slotPitchChanged(double pitch)`.

I pushed an experimental commit, let's see if that's sufficient.

ronso0 replied:
> I pushed an experimental commit, let's see if that's sufficient.

No, it's not.
Can we use NEAR or similiar?

Swiftb0y replied:
probably, that's what `TrackLoadResetsPitch` already does anyways.

https://github.com/mixxxdj/mixxx/blob/0119d85562a81976c3c7678d58cecfae4d0adc5a/src/test/enginebuffertest.cpp#L52-L56

ronso0 replied:
:facepalm:

ronso0 committed:
DEBUG (Remove!): add pitch_adjust slider to LateNight

ronso0 commented:
Alright, `ASSERT_NEAR` makes the tests succeed.
Ready to roll!

@daschuer wanna take a final look?

ronso0 committed:
KeyControl: std::unique_ptr, parented_ptr

ywwg commented on src/engine/controls/keycontrol.cpp:
```diff
@@ -234,15 +246,8 @@ void KeyControl::updateRate() {
-                // If 'current' aka 'not original' key was locked
-                // TODO(ronso0) Why does it depend on keylock mode?
-                if (m_keylockMode->get() == kLockCurrentKey) {
-                    // reset to linear pitch
```
it's not that we reset to linear pitch, it's that we allow rate slider changes to affect the musical pitch.  But 0.0 rate adjustment may no longer the original key.

ywwg commented:
I wrote a lot of text that was all wrong -- I think this fix is correct! thank you <3

ronso0 commented:
🎉
